### PR TITLE
Dream integration

### PR DESCRIPTION
**Pull request by jgalan**
Large jgalan  [](https://gitlab.cern.ch/rest-for-physics/rawlib/-/commits/dream_integration) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=rest-for-physics&utm_campaign=badge_info)

Recovering dream electronics from existing REST v2.2 code.

### CODE REVIEW

**Comment by jgalan** [comment]
```
                                   ====================================================================================================                                   
                                   ||                             Process : TRestRawDreamToSignalProcess                             ||                                   
                                   ||                  Name: virtualDAQ  Title: defaultTitle  VerboseLevel: warning                  ||                                   
                                   ||                        -----------------------------------------------                         ||                                   
                                   ||                                                                                                ||                                   
                                   ||                                                                                                ||                                   
                                   ||                              ====================================                              ||                                   
                                   ||                                       DAQ : defaultTitle                                       ||                                   
                                   ||                                    Electronics type : Dream                                    ||                                   
                                   ||                                 Minimum number of points : 200                                 ||                                   
                                   ||                              All raw files open at beginning : 1                               ||                                   
                                   ||                              ====================================                              ||                                   
                                   ||                                                                                                ||                                   
                                   ||                                                                                                ||                                   
                                   ====================================================================================================                                   
                                                                      
                                   TRestProcessRunner: 1 + 0 processes loaded, 1 threads prepared!                                   

                                   ====================================================================================================                                   
                                   ||                             Process : TRestRawDreamToSignalProcess                             ||                                   
                                   ||                  Name: virtualDAQ  Title: defaultTitle  VerboseLevel: warning                  ||                                   
                                   ||                        -----------------------------------------------                         ||                                   
                                   ||                                                                                                ||                                   
                                   ||                                                                                                ||                                   
                                   ||                              ====================================                              ||                                   
                                   ||                                       DAQ : defaultTitle                                       ||                                   
                                   ||                                    Electronics type : Dream                                    ||                                   
                                   ||                                 Minimum number of points : 200                                 ||                                   
                                   ||                              All raw files open at beginning : 0                               ||                                   
                                   ||                              ====================================                              ||                                   
                                   ||                                                                                                ||                                   
                                   ||                                                                                                ||                                   
                                   ====================================================================================================                                   
                                                                      
                                   ====================================================================================================                                   
                                   [==                          TRestProcessRunner: Starting the Process..                          ==]                                   
TRestRawDreamToSignalProcess::ProcessEvent: 100 events processed in file, and 0 bad events skipped ----------------------------------------------------------------------]
TRestRawDreamToSignalProcess::ProcessEvent: 200 events processed in file, and 0 bad events skipped ----------------------------------------------------------------------]
TRestRawDreamToSignalProcess::ProcessEvent: 300 events processed in file, and 0 bad events skipped ----------------------------------------------------------------------]
TRestRawDreamToSignalProcess::ProcessEvent: 400 events processed in file, and 0 bad events skipped ----------------------------------------------------------------------]
TRestRawDreamToSignalProcess::ProcessEvent: 500 events processed in file, and 0 bad events skipped ----------------------------------------------------------------------]
```

This is a first attempt to migrate existing dream process data previously used in REST v2.2 to REST v2.3. The process seems to be reading the events properly, although progress bar seems not moving as usual.

Files are placed at https://sultan.unizar.es/exchange/

`fdf` raw file, and `01_raw.rml`, `run.xml` and `globals.xml`

**Comment by jgalan** [comment]
After long time processing, I get the following warning message

```
TRestRawDreamToSignalProcess::ProcessEvent: 4200 events processed in file, and 0 bad events skipped ---------------------------------------------------------------------]
-- Warning : TRestRawDreamToSignalProcess::ReadFeuHeaders: Problem in reading raw file, ferror 0 feof 1 fInputBinFile 0x5649a9d64560
-- Warning : TRestRawDreamToSignalProcess::ProcessEvent: Error in reading feu header (bad file or end of file), trying to go to the next file
                                   Waiting for processes to finish ...                                   
                                   [==                          TRestProcessRunner: 4247 processed events                           ==]                                   
                                   Configuring output file, merging thread files together                                   
                                   [==                                     TRestRun Created ...                                     ==]                                   
                                   [==                                 - Path : /home/jgalan/data/                                  ==]                                   
                                   [==                - Filename : R-0001_00000_RawData_preserve_jgalan_V2.3.5.root                 ==]                                   
                                   [==                                            Done!                                             ==]                                   
```

**Comment by barbarabias** [comment]
I don't know if it could be useful, but I add here some additional information provided by Damien when he shared the files:

"[...] it works well to read data taken
without zero suppression (with zero suppression I have not really
tested). It is even able to read several raw files one after the other
(for instance with restManager --i '/any/path/data_files_*.fdf'), but it
does not work if you want to merge data from several FEU cards in
parallel [...]. Here is an example of the entry in the config.rml file to activate this process:

     <addProcess type="TRestFEUDreamToSignalProcess" name="virtualDAQ"
value="ON" verboseLevel="warning"  >
       <parameter name="electronics" value="FEUDream"/>
       <parameter name="pedScript" value="ped"/>
       <parameter name="runScript" value="run"/>
       <parameter name="minPoints" value="200" />
       <parameter name="showSamples" value="255" />
       <parameter name="fileFormat" value="FEU" />
     </addProcess>

Most of the parameters are for TRestRawToSignalProcess or useless, but
minPoints should be equal to the number of samples which are settled in
the FEU configuration."

**Comment by nkx111** [comment]
> The process seems to be reading the events properly, although progress bar seems not moving as usual.

To make the progress bar working we need to increase`TRestRawToSignalProcess::totalBytesReaded` each time we read the file for certain bytes. The unit is bytes.

```
fread(fHeaderFrame[i].frameHeader, 256, 1, fInputFiles[i]);
totalBytesReaded += 256;
```

**Comment by nkx111** [comment]
> There is also something strnage about the initialization of All raw files open at beginning : 1 which is initialized to 0 from config file when it reads "Dream" electronics. Perhaps thats not a big issue?

In the current external processes we indeed open all the files at the beginning. Even if the daq is still writing data, it is still OK. I don't know why we need to open the files one by one in dream system.

> but it does not work if you want to merge data from several FEU cards in parallel

Merging data from several cards is possible for CoBo system. Maybe you can look into `TRestRawMultiCoBoAsAdToSignalProcess.cxx` to have some hints?

**Comment by jgalan** [comment]
Ok I added `totalBytesReaded` however for some reason the estimation is not right. When it finishes with 4000 events tells is about 0.3%.

There is the warning at the end, but still I believe the resulting data output is probably good.

Screenshot 2021-05-26 at 10 58 29

**Comment by nkx111** [comment]
> Ok I added totalBytesReaded however for some reason the estimation is not right. When it finishes with 4000 events tells is about 0.3%.

You should put the lines outside if conditions!

```
int a = fread((char*)&(Feu.current_data), sizeof(Feu.current_data), 1, fInputBinFile);
        totalBytesReaded += sizeof(Feu.current_data);
        if (a == 0) {
            return true;  // failed
        }
```

https://github.com/rest-for-physics/rawlib/blob/bdff7f9cfb1dd48179573fbd7f69c01bb6263ab3/src/TRestRawDreamToSignalProcess.cxx#L307

**Comment by jgalan** [comment]
Ups, thats a noob error xD. Fixed now.

Screenshot 2021-05-28 at 13 16 21

Thanks!

**Comment by nkx111** [comment]
The reading speed seems to be too slow. Optimization is still needed. Usually the reading speed shall > 20MB/s.

**Comment by jgalan** [comment]
Right, I agree it needs optimisation. I may just comment on the documentation as a TODO.

Multi-threading will not work here because it is an external process? If the input binary would be a ROOT file we would profit from multithreading?

**Comment by nkx111** [comment]
> If the input binary would be a ROOT file we would profit from multithreading?

No. External processes are single threaded.

**Comment by jgalan** [comment]
Right, but if the ROOT file would contain, an `eventTree`, an `analysisTree`, and a `TRestRun` then, it would not be external anymore, and it would be able to profit, right? I mean, if the DAQ produces such file @juanangp

**Comment by juanangp** [comment]
> Right, but if the ROOT file would contain, an `eventTree`, an `analysisTree`, and a `TRestRun` then, it would not be external anymore, and it would be able to profit, right? @nkx111 I mean, if the DAQ produces such file @juanangp

Yes, in principle would be possible to save the data in a root file with TRESTRawSignalEvent format. I am working on a DAQ software to be included in REST, but it would require time and a lot of work...

**Comment by jgalan** [comment]
> > Right, but if the ROOT file would contain, an `eventTree`, an `analysisTree`, and a `TRestRun` then, it would not be external anymore, and it would be able to profit, right? @nkx111 I mean, if the DAQ produces such file @juanangp
> 
> Yes, in principle would be possible to save the data in a root file with TRESTRawSignalEvent format. I am working on a DAQ software to be included in REST, but it would require time and a lot of work...

Sure, I know, I just want to make sure we will be able to take advantage of multi-threading in that scenario. I guess it is the case, right? @nkx111

File: inc/TRestRawDreamToSignalProcess.h [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        bool is_Feu_header() const { return (((data)&0x7000) >> 12) == 6; }      // X110
+        bool is_data_header() const { return (((data)&0x6000) >> 13) == 1; }     // X01X
+        bool get_zs_mode() const { return (((data)&0x400) >> 10); }
+        int get_Feu_ID() const { return (((data)&0xFF)); }
+        long int get_finetstp() const { return (((data)&0x0007)); }
+        //#define GET_FINETSTP(word)      (word & 0x0007)
+        int get_sample_ID() const { return (((data)&0xFF8) >> 3); }
+        int get_channel_ID() const { return (((data)&0x3F)); }
+        int get_dream_ID() const { return (((data)&0xE00) >> 9); }  // non-zS mode
+        int get_dream_iD() const { return (((data)&0xE00) >> 6); }  // for zS mode
```
Comment (juanangp) [inline inc/TRestRawDreamToSignalProcess.h:78]: The difference between upper and lower case in `get_dream_iD()` is error prone. I would suggest to do `get_dream_iD( bool zS)`

File: inc/TRestRawDreamToSignalProcess.h [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        long int Cell_ID;
+        int Cell_ID_ISB;
+        int Cell_ID_LSB;
+        int Cell_ID_MSB;
+        long long int TimeStamp;
+        long int TimeStamp_Op1;
+        long int TimeStamp_Op2;
+        long int TimeStamp_Op3;
+        long int FineTimeStamp;
+        int isample;
```
Comment (juanangp) [inline inc/TRestRawDreamToSignalProcess.h:116]: Same comment as before, please consider to move to `uint8_t`, `uint16_t`, `uint32_t` or `uint64_t` instead.

File: src/TRestRawDreamToSignalProcess.cxx [hunk 3]
```diff
@@ -0,0 +1,587 @@
...
+            return NULL;
+        }
+
+        debug << "TRestRawDreamToSignalProcess::ProcessEvent: returning signal event fSignalEvent "
+              << fSignalEvent << endl;
+        if (GetVerboseLevel() > REST_Debug) fSignalEvent->PrintEvent();
+        return fSignalEvent;
+    }
+
+    return NULL;  // can't read data
```
Comment (juanangp) [inline src/TRestRawDreamToSignalProcess.cxx:179]: `return nullptr`

File: src/TRestRawDreamToSignalProcess.cxx [hunk 4]
```diff
@@ -0,0 +1,587 @@
...
+    fSignalEvent->SetID(Feu.EventID);
+    fSignalEvent->SetTime(tStart + Feu.TimeStamp * 8.E-9);  // timeStamp in seconds
+
+    return false;
+}
+
+bool TRestRawDreamToSignalProcess::ReadDreamData(FeuReadOut& Feu) {
+    register bool got_raw_data_header = false;
+    register bool got_channel_id = false;
+    register int ichannel = 0;
```
Comment (juanangp) [inline src/TRestRawDreamToSignalProcess.cxx:326]: `register` seems deplecated
https://stackoverflow.com/questions/3207018/register-keyword-in-c

Please consider to remove it

Comment (jgalan) [reply src/TRestRawDreamToSignalProcess.cxx:326]: So, what we should do? just remove it?

Comment (juanangp) [reply src/TRestRawDreamToSignalProcess.cxx:326]: I think so
`register bool got_raw_data_header` --> ` bool got_raw_data_header`

File: inc/TRestRawDreamToSignalProcess.h (hunk 2 above)
Comment (jgalan) [reply inc/TRestRawDreamToSignalProcess.h]: Ok, it happened, changing `int` to `uint32_t`, `short int` to `uint16_t` and `long int` to `uint64_t` provoked the validation to fail.

When I reverted the change, the pipeline succeeded again.

https://gitlab.cern.ch/rest-for-physics/rawlib/-/commit/bc9b1614369914d6c81dec76ae96c12e3629e398

It must be a delicate change...

Comment (juanangp) [reply inc/TRestRawDreamToSignalProcess.h]: Perhaps the problem is the signature? But in general shouldn't be an issue, I did similar changes in #12 and I didn't saw any issue, but I don't know if there is validation...

Can you just create an issue about this to be followed up?

**Comment by nkx111** [comment]
> Right, but if the ROOT file would contain, an eventTree, an analysisTree, and a TRestRun then, it would not be external anymore, and it would be able to profit, right?

You mean in case that raw data is directly provided in ROOT file by daq? No, it would still be single threaded. The **reading** of the events from file is always one by one.

**Review by barbarabias — APPROVED** [review]

**Review by nkx111 — APPROVED** [review]

**Comment by ddn2** [comment]
> In the current external processes we indeed open all the files at the beginning. Even if the daq is writing data, it is OK to read them(although the progress bar will exceed 100% in this case). I don't know why we need to open the files one by one in Dream daq system.

Well I needed to have a mechanism to read several raw data files in a raw, but I seen the existing code in TRestRawToSignalProcess was opening all the raw files in the same time at the beginning of the job. I preferred to implement an optional behavior where raw files are opened and closed one after the other, in order to limit the number of opened files in linux.  It's optional because I realized that to have all the files opened was needed by the MultiCoBo process.

> > but it does not work if you want to merge data from several FEU cards in parallel
> 
> Merging data from several cards is possible for CoBo system. Maybe you can look into `TRestRawMultiCoBoAsAdToSignalProcess.cxx` to get some hints?

Actually I wrote a long time ago a standalone software to decode FEU Dream raw data (a large part of the code of TRestFEUDreamToSignalProcess comes from it), which featured the readout of several parallel card files. I plan to reimplement that feature to the REST process, but I don't know when I will have time to do that...

**Pull request merged**